### ISSUE TITLE
3 character ICD-10 codes have `X` appended to them in SUS tables

### ISSUE THREAD

**Issue by alexwalkerepi** [opensafely-core/cohort-extractor#512]
The majority of ICD-10 codes are 4 characters, e.g. `I676`, but some are 3 characters, e.g. `G08`. In the SUS data, it appears that there are no recorded  3 char codes, but instead, they all have `X` appended to them: `G08X`

This appears to be true at least for the `APCS` table, but I haven't checked other SUS tables.

*However* we can't just append `X` to all 3 char ICD-10 codelists, because in the `ONS_Deaths` table has 3 char codes as you'd expect.

**Comment by evansd** [comment]
I _think_ the ideal answer here is for the cohortextractor to automatically pad 3 (or fewer?) character ICD10 codes with `X`s for queries on the `APCS` table. And then automatically strip them when returning codes. So from the point of view of the user , it's as if the `APCS` data was never `X`-padded in the first place. The idea being that this is an implementation detail of the way `APCS` happens to be encoded which we abstract away from the user entirely.

I'm not sure yet how fiddly this will be to implement, but does that sound broadly sensible as an approach?

**Comment by krishnanbhaskaran** [comment]
I think that would work for the Xs., but there's another issue mentioned in the Slack thread... in some cases code lists may say J12 when they really mean J12*.

**Comment by krishnanbhaskaran** [comment]
Yes I think that would work for handling the Xs and would mean the user essentially doesn't have to worry about it.

**Comment by alexwalkercebm** [comment]
Yes I agree that approach should work. I don't *think* there are actual ICD-10 codes with fewer than 3 characters, but perhaps worth bearing #382 about hierarchical categorisation in mind while looking.

**Comment by evansd** [comment]
Just looking at how this is currently implemented I think there's another issue here.

If you use `with_these_primary_diagnoses` it will do an exact match for codes, meaning that any with the `X` suffix won't match and that codes like `J12` won't match decedents like `J120`, as discussed above.

But if you use `with_these_diagnoses` then it will use prefix matching meaning that e.g. `G08` **will** match `G08X` and `J12` **will** match `J120`.

This is definitely not the behaviour we want, but I need to look at what studies this might effect before changing it.

**Comment by alexwalkercebm** [comment]
Yes we definitely don't want that behaviour.

It's possible this will slightly affect [this repo](https://github.com/opensafely/post-covid-outcomes-research), which I think just uses `with_these_diagnoses`, but uses codelists with 3 character codes, so it may be picking up unexpected additional codes that are children of the 3 character ones.

**Comment by krishnanbhaskaran** [comment]
Thanks Dave - agree that the two should work in a consistent way, and I think ideally you'd want it to include the specified codes only by default, but to have a way of using a wildcard (e.g. by putting a wildcard character in the code list, e.g. J22*). 

In this repo (https://github.com/opensafely/post-admission-admissions-research ) I've mainly used with_these_primary_diagnoses. For now I am using full code lists (with all children specified) and have appended X to the 3-char ones and saved locally - so I think that should be ok as a workaround for now? I will then update appropriately once this issue is resolved and the function updated.

**Comment by evansd** [comment]
@inglesp It looks like the prefix matching thing was actually intentional because in "fixing" I've discovered a test which asserts exactly that behaviour:
https://github.com/opensafely-core/cohort-extractor/blob/a87a25c85d58d416a401ee763030ce772bd4c942/tests/test_tpp_backend.py#L2853-L2857

(There are no codes in the test fixture matching `XXX` but there are several which _start_ `XXX`)

I wonder if an alternative way of fixing this is to actually do the reverse of what we were originally planning and change the behaviour for `with_these_primary_diagnoses` to automatically match on prefixes as well.

@alexwalkercebm @krishnanbhaskaran Are there any circumstances where you _wouldn't_ want a three char ICD-10 code to match its 4 char child codes?

**Comment by alexwalkercebm** [comment]
I think in a large majority of cases you would want to match the child codes, because when you use a 3 char parent code it's sort of implied that it is a more general term that encompasses all of the concepts within the children. I've had a quick look through some of the ICD codelists for [this study](https://github.com/opensafely/post-covid-outcomes-research) and haven't found any examples where we've included a parent but excluded one of it's children.

However I can imagine there could be edge cases. The only practical use that I can think of offhand is if you were just looking very specifically at the number of people that were coded using the three character code, compared to a 4 char code.

If we went for that option we'd need to make it very clear that this is what happens, and ideally find a better option later.

**Comment by krishnanbhaskaran** [comment]
I can't think of any situations where I have needed 3-char codes only but not their children - generally if you specify 3-char either (a) there aren't any children in ICD10 for that code, or (b) as Alex says, you are specifying at a broader level so would want all children.  

Note that in general when an ICD10 code has 4th-char children in the dictionary, then the 3-char code doesn't have a separate meaning, it just really means all diagnoses in that category. So the behaviour you suggest makes total sense. And if you only wanted a subset of the children you would just specify at 4-char level.

So I would say yes; making with_these_primary_diagnoses behave in the same way (automatically picking up children of 3-char codes) would seem fine to me.

**Comment by brianmackenna** [comment]
Not sure of the correct etiquette but sorry to reopen if thats not the right thing!

I can't find any of this in the documentation. While trying to fix [this issue](https://github.com/opensafely-core/opencodelists/issues/1075) I wasn't understanding it as these codelists seem to behave in a different way to other codelists.

From discussion above I think this means that all children are included regardless of a parent. I expected that no children were included like all other codelists so couldn't work out why are results were so high. We then looked at this and realised we actually had an edge case where we wanted to exclude a child. Due to another error we no longer need to do this - so I think no need to change implementation but should this behaviour be more explicit in the docs?

**Comment by alexwalkerepi** [comment]
The different implementation is caused by the very structured hierarchy of ICD-10, which enables a different way that `cohortextractor` queries codes for the SUS data. This has the advantage of just being able to query higher levels of the hierarchy without having to explicitly list the codes.

Certainly worth considering whether a feature to include e.g. `Z94` and all children except `Z94.1`.

**Comment by HelenCEBM** [comment]
Snippet from [docs](https://docs.opensafely.org/study-def-variables/#cohortextractor.patients.admitted_to_hospital) - it's mentioned that diagnosis code searches work this way but not specified for procedures 
![image](https://user-images.githubusercontent.com/24392156/153590594-6026b6f8-24cc-45ab-bb01-b3ab9be4a041.png)

Can we clarify this? Do we know if there are Xs in short procedure codes too?

**Comment by HelenCEBM** [comment]
> Certainly worth considering whether a feature to include e.g. Z94 and all children except Z94.1.

This could be useful but could work in one of two ways, e.g. let's say a patient has both `Z941` and `Z942` in a single record Do we want to **include** this record because of the presence of `Z942` or **exclude** it because of the presence of `Z941`? Might need to allow a user to specify whether the inclusion list or exclusion list is more dominant.
